v3/lints/cabf_br/lint_sub_ca_eku_incompatible_values.go

```diff
@@ -0,0 +1,94 @@
+/*
+ * ZLint Copyright 2021 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/*
+ * This lint contributed by Adriano Santoni <adriano.santoni@staff.aruba.it>
+ * of ACTALIS S.p.A. (an ARUBA company). Last revised June 30, 2021.
+ *
+ * Checks that the EKU extension in a Subordinate CA certificate meets the
+ * "MUST NOT" requirement set out in the BRs 7.1.2.2, letter g):
+ *
+ *     For Subordinate CA Certificates that will be used to issue
+ *     TLS certificates, the value id-kp-serverAuth[RFC5280] MUST be present.
+ *     The value id-kp-clientAuth[RFC5280] MAY be present.
+ *     The values id-kp-emailProtection[RFC5280], id-kp-codeSigning[RFC5280],
+ *     id-kp-timeStamping[RFC5280], and anyExtendedKeyUsage[RFC5280] MUST NOT
+ *     be present. Other values SHOULD NOT be present.
+ *     For Subordinate CA Certificates that are not used to issue TLS certificates,
+ *     then the value id-kp-serverAuth[RFC5280] MUST NOT be present.
+ *     Other values MAY be present, but SHOULD NOT combine multiple independent usages
+ *     (e.g.including id-kp-timeStamping[RFC5280] with id-kp-codeSigning[RFC5280]).
+ */
+
+package cabf_br
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_sub_ca_eku_incompatible_values",
+		Description:   "Subordinate CA extkeyUsage: if serverAuth is present, then emailProtection, codeSigning, timeStamping, and anyExtendedKeyUsage MUST NOT be present.",
```

Comment:
So this is tricky, because 7.1.2.2 contains the following language:

> For Cross Certificates that share a Subject Distinguished Name and Subject Public Key with a Root Certificate operated in accordance with these Requirements, this extension MAY be present. If present, this extension SHOULD NOT be marked critical. This extension MUST only contain usages for which the issuing CA has
verified the Cross Certificate is authorized to assert. This extension MAY contain the anyExtendedKeyUsage [RFC5280] usage, if the Root Certificate(s) associated with this Cross Certificate are operated by the same organization as the issuing Root Certificate.

As is, this lint would fire an Error for those certificates, due to lacking sufficient context about whether or not the given certificate being linted meets that language, right?

Should this lint be blocked on https://github.com/zmap/zlint/issues/592 (see also https://github.com/zmap/zlint/issues/581#issuecomment-809529934 for more discussion about that issue)

Reply:
@sleevi  So I have a branch up at https://github.com/zmap/zlint/tree/context that at least gets us the idea of what we want.

Take the following lint into consideration.


```go
type MyLint struct {
	BottlesOfBeerOnTheWall int
}

func init() {
	lint.RegisterLint(&lint.Lint{
		Name:          "e_ca_too_few_beers",
		Description:   "CA Certificates MUST have at least 99 beers.",
		Citation:      "BRs: 7.1.4.3.1",
		Source:        lint.CABFBaselineRequirements,
		EffectiveDate: util.CABV148Date,
		Lint:          NewMyLint,
	})
}

// This is the new thing. You give us a struct to deserialize
// into and we will get the appropriate context into it.
//
// In this case, the lint itself holds the data in quesiton.
func (l *MyLint) Configure() interface{} {
	return l
}

func NewCaCommonNameMissing2() lint.LintInterface {
	return &MyLint{}
}

func (l *MyLint) CheckApplies(c *x509.Certificate) bool {
	return util.IsCACert(c)
}

func (l *MyLint) Execute(c *x509.Certificate) *lint.LintResult {
	if l.BottlesOfBeerOnTheWall < 99 {
		return &lint.LintResult{Status: lint.Error}
	} else {
		return &lint.LintResult{Status: lint.Pass}
	}
}
```

From there we can build the additional context file...

```bash
$ cat >> ctx.toml <<EOF
[e_ca_too_few_beers]
BottlesOfBeerOnTheWall = 98

[e_some_other_lint]
is_web_cert = false
EOF
```

...and shove it into ZLint

```bash
$ ./zlint -nameFilter e_ca_too_few_beers -context ctx.toml cert.pem
{"e_ca_too_few_beers":{"result":"error","details":"Time for a beer run!"}}
```

Of course it needs more testing/user acceptance testing and it needs integrations into the large test corpus.

Additionally, the framework only configures on a per lint basis. That is, each section in TOML document lines up to a specific lint, however I am not passing in any sort of global context to all lints. This is only because I am apprehensive about either maintaining a well defined global struct _or_ passing anonymous `map[string]interface{}` to every lint (because that just sounds chaotic).